### ISSUE TITLE
more robust/specific support for background execution

### ISSUE THREAD

**Issue by ethael** [flutter/flutter#24278]
@eseidel  encouraged people in #3671, that they should create proposals on what is missing in the current background execution implementation and how to make it ready enough for wide range of usages and people skills. So I decided to write down my view on it.

People tend to use background execution mostly for 3 main purposes:

**1. Do some work out of main UI thread when application is in the foreground**
- This is already mostly covered by basic dart language async features. 
- What is missing is having an ability to ensure (platform agnostic), that "async code" will finish even if app goes to background. Flutter could provide simple API to wrap custom async code into, that would add such functionality.
- native support
  - iOS supports it using: [Executing a Finite-Length Task](https://apple.co/2B2GP2p) feature
  - Android support is by default afaik

**2. Do something on the background periodically**
- Android recently dropped support classic background services using custom separate thread implementation in favour of scheduling execution together with various limitations applied based on actual device state. This is basically the standard iOS way. 
- Flutter does not have any out of the box support for this. There is a 3rd party android_alarm_manager plugin created by flutter team itself, which is obviously android specific. Couple of days ago, @christocracy ported his cordova/react-native [background_fetch plugin](https://pub.dartlang.org/packages/background_fetch) to flutter that basically does what most people waited for. But it is still necessary to be made by flutter team to the flutter core. This needs to be maintained by Google/Flutter itself to ensure such important feature does not stand on one man's shoulders.
- native support
  - iOS supports it using "background fetch"
  - Android supports it using AlarmManager and JobScheduler

**3. Trigger action based on listening on some platform event**
- both platforms have ability to register for location change, remote notification arrival and bluetooth events. android has many more and current "best practice" in Android is registering programmatically for receiving events of specific type using BroadcastReceiver.
- flutter has its own robust support for firebase related features that includes also remote notification arrival handling. @christocracy has released his port of [background geolocation plugin](https://pub.dartlang.org/packages/flutter_background_geolocation). But there is no standard API to register for some kind of event and ability to consume it in the background. This is not trivial but also not too hardcore task to be landed. Common Flutter API could have ability to register for any event defined by set of "event type" constants. native implementation will then provide support for those types. This API may have ability to implement and register custom event types so plugin developers would be able to add other event implementations but registering and consuming by end programmers would be done through the same unified Flutter API.
- another very important reason why this has to be provided by flutter itself is, that on iOS platform, you can register only one delegate per app for receiving events. (for example notifications) and therefore you currently can not use [local notifications plugin](https://pub.dartlang.org/packages/flutter_local_notifications) together with [firebase messaging plugin](https://pub.dartlang.org/packages/firebase_messaging). It is explained in detail at the end of [readme file](https://github.com/MaikuB/flutter_local_notifications/blob/master/README.md) of the local notifications plugin. So if the registering delegate would be flutter, which would provide ability to register on those events in dart layer, all plugins would be able to run next to each other and won't block themselves.

**TL;DR**
So the concrete changes I would like this issue to resolve into are:

**1.** Implement Flutter API + native counterparts that will ensure complete finish of a foreground async method even if the app will go to background

**2.** Implement Flutter API + native counterparts for executing background tasks every N minutes

**3.** Implement Flutter API + native counterparts for listening to platform events such as location change, push notifications, OS boot finish, and other

**Comment by radvansky-tomas** [comment]
Thank you for your contribution, I would like to add some suggestions about implementation of such abstraction as I think that for any developer trying to use flutter is initial template, tools and documentation the main pattern what they will use in their future flutter project. Therefore I suggest to implement "headless" callback in every flutter main.dart (entry point). And then leave option to implement headless code to plugin developers.

**Comment by eseidelGoogle** [comment]
Thanks @ethael!  Could you edit the description to highlight the parts you would like us to take action on?  e.g. the concrete changes you would like this issue to resolve into?  Thanks!

FYI @bkonyi

**Comment by MaikuB** [comment]
Thanks for the write-up @ethael. Just wanted to point out I've actually already created another issue that is covered in the 3rd point around delegates already https://github.com/flutter/flutter/issues/22099

**Comment by ethael** [comment]
@eseidelGoogle done. I added TL;DR section. Numbers correspond to detailed numbered sections above it

**Comment by eseidelGoogle** [comment]
@bkonyi (or @cbracken) can you comment on each of the status/feasibility of 1, 2 and 3 above?  Thanks!

**Comment by christocracy** [comment]
@bkonyi 
> If @christocracy is willing, I'm sure the Flutter team would be happy to accept the background_fetch plugin as an open source contribution to the Flutter project

I have no problem with that.  However, while the plugin is *very* simple, the core iOS and Android libs are packed into a `.framework` and `aar` respectively, built from a separate project [`transistor-background-fetch`](https://github.com/transistorsoft/transistor-background-fetch/blob/master/android/tsbackgroundfetch/build.gradle).  I use the generated `.framework` and `aar` in a number of other frameworks (Cordova, React Native, NativeScript, pure native apps and now Flutter).  And one of my other libs [`flutter_background_geolocation`](https://pub.dartlang.org/packages/flutter_background_geolocation/versions/0.2.0#-analysis-tab-) *requires* `background_fetch` as a dependency, particularly the existence of the `TSBackgroundFetch.framework`.

**Comment by bkonyi** [comment]
@ethael #16662 should be resolved once the change in flutter/engine#6858 is rolled into the framework. If you know of any other situations where only a single plugin can register a delegate feel free to file an issue against me and I'll get it fixed :)

@christocracy glad to hear you're open to the idea! I'll take a closer look at the plugin to see if that's something we would like to accept as a first party plugin. @cbracken @bparrishMines, would you mind taking a look also? If not, feel free to direct this to someone who might have more background.

**Comment by ethael** [comment]
@bkonyi thank you for quick response. I am not aware of any other registering collision right now. The only other issue regarding these topics is #23935. But to the main points:

1. From your reaction to point 1, it looks for me like you want to do it for IOS only. Please keep platform decision making on the API side. The single biggest purpose of hybrid framework is to minimize need for platform specific conditions. So when android already does what we need, good for us, but API should be platform agnostic. Devs just want to use `runEvenIfAppGoesToBackground(someLongMethod);` and not checking if his app is deployed on IOS first. 

2. Here I have the same issue. It does not really matter for most people, that android_alarm_manager is able to cover most use cases on Android when their app should run on both platforms. I am happy that @christocracy is willing to donate his work. Hope you wouldn't mind that his approach uses same isolate. Most people I spoke about this topic were happy that it does not use separate isolate because  they don't need to handle problems with multiple singleton instances, or not very well functioning SharedPreferences plugin, as in case of  android_alarm_manager, it is instantiated once in main thread and once in the isolate and changes in one instance are not visible in the other. Maybe using separate isolate can be optional, if needed.

3. Maybe you don't need to implement and maintain all hooks. Create generic API that can be extended by registering custom event handlers. Then implement only location events by yourselves. Thanks to this, most important feature will be landed and people will also have blueprint on how to do it for another type of event. Then let the community do the rest. But create unified API for end devs that will make using flutter even one step faster and more convenient.

Are these approaches feasible @eseidel @bkonyi?

**Comment by baneageorge** [comment]
Hi guys, I really new to flutter.
I am considering switching my app from Xamarin and I am evaluating how easy would it be to achieve it in Flutter.

One scenarion that would be good to cover:
I have a mobile app that connects to a bluetooth device, this is not bindable by the operating system.
I need a process/task to run in the backround to check when the device is in range, let's say every 10-15 seconds and another process that manages the communication which requests data from the bluetooth dongle every 200-500ms.

Curently I am able to achieve this very easly in Xamarin by following this tutorial: https://robgibbens.com/backgrounding-with-xamarin-forms/
Maybe you can think of a similar solution.
Basically, for each background task you have to generate 1 file for each platform one shared file that is platform agnostic and contain all the logic. The comunication between them is done through eventbus mechanism.

**Comment by matthew-carroll** [comment]
@baneageorge the way I would approach this in Flutter is to write my own plugin.  Not a generic Bluetooth plugin, but a plugin that exposes only what my app needs with regard to Bluetooth communication.  This way you reserve the ability to do anything that the OS needs you to do.  For example, the background execution limitations on Android are much different than iOS.  You may not always be able to abstract these behaviors.  Therefore, I would implement my own Service on Android to do this, as well as my own analog on iOS.  Then I would expose those systems to my Flutter app as a plugin.  This process isn't the easiest process, but most of the complexity is due to Android and iOS background execution concerns rather than Flutter APIs.

**Comment by kkarakk** [comment]
@matthew-carroll  yes but the point is to abstract away those concerns so flutter truly becomes an alternative to just rolling your own native  iOS and Android app. at this point,if you've got significant background task based functionality and a small-ish team - flutter just becomes overhead instead of helping to reduce workloads. why would i not choose xam forms/react instead of flutter? 
backgrounding is increasingly becoming a hot topic issue as android vendors have started to aggressive throttle most methods of doing it and the research involved is not trivial. i too have decided against switching the team from xamarin to flutter until the backgrounding issue is sorted out.

**Comment by matthew-carroll** [comment]
@kkarakk Background execution is a topic that is dramatically different than the specific notion of doing Bluetooth work.  Bluetooth, especially BLE behavior, is highly specific.  Yes, Flutter requires much stronger support for generic background work *in Dart*, but to conflate that with how one should effectively program for Bluetooth/BLE would be a mistake.  There is no abstraction that will unify BLE capability on Android and iOS, nor is it likely a good use of Flutter development resources to continuously update such an abstraction.  That said, it is possible that others on the Flutter team feel differently and may choose to provide something for Bluetooth.  But I probably wouldn't, and I would definitely separate Bluetooth/BLE from the larger topic of "how do I execute Dart code in the background."

If you believe you have specific examples or use-cases that illustrate how Flutter would conceivably treat Bluetooth the same way on iOS and Android, I'm more than happy to take a look at that information.

**Comment by radvansky-tomas** [comment]
Sure @matthew-carroll , but this topic is about generic background support. iOS and Android has different but unified approach how to tread background tasks. And this is something which flutter should support. I mean whole point of any cross platform framework is to share business logic between those platform. That said, yes @baneageorge has to create bluetooth communication bits for both platform, BUT currently he cannot call flutter to perform logic from that background task. Things like update DB, access models, math functions etc

**Comment by baneageorge** [comment]
Sorry guys, I think I created some misunderstanding.
There is already a cross-platform Bluetooth library for flutter, right? (https://github.com/pauldemarco/flutter_blue)
What I wanted to say, it is that I need continuous running processes in the background to keep the connection open with the Bluetooth device when the user switches to some other app or turns off the screen. (I am not sure this case is covered in the 3 main purposes from the top of this issue)

**Comment by matthew-carroll** [comment]
@baneageorge there are 2 perspectives on this issue:

1. A developer wants to write Dart code to control Bluetooth.  In that case, yes you would need some background execution support for Dart/Flutter that is currently not available.  We are working on figuring out how things like that should be done in Flutter and something will eventually be available.

2. A developer writes all Bluetooth code on the platform side and then uses method channels to send the appropriate information over to Flutter whenever Flutter wants it.

My recommendation was to go with option 2.  Not only is option 2 something that you can do right now, but I honestly think it would be simpler in the long-run because BLE has somewhat different APIs between Android and iOS, and you always need to take into consideration the realities of the platform hardware.  For example, a few years ago when Android 5 was new, and Android 4.x was still popular, we had to manually activate and deactivate the Bluetooth radio on Android to keep the scanner working.  This was a bug in the OS.  Not only was it specific to Android, but it was specific to a particular version of Android.  A Dart abstraction layer will never cover these edge cases.

**Comment by rmawatson** [comment]
I'm using flutter_blue, and have it working in a separate service on Android that launches its own isolate. So it still runs when the activity is killed. On iOS with some modifications to the plugin it will run when the app is backgrounded and and has blue-central role. 

However, on iOS if there are current connections, pending connections or ongoing scans for specific BLE service's and if the app is killed it will be automatically restarted when a scan finds something. iOS then calls centralManager:willRestoreState on the CBCentralManagerDelegate.

The problem here is it flutter will also start launch the main isolate and run any code in main(). When the app is invoked using willRestoreState it has about 10 seconds to complete any work. What I really want is to be able to only launch my isolate, run my callback to allow it to be handled in dart, but without invoking the main isolate.

**Comment by nerdyadventurer** [comment]
> Trigger action based on listening on some platform event

This is something I want to in order to choose Flutter for an app? Any plans to support this?

**Comment by matthew-carroll** [comment]
@nerdyadventurer technically that is already supported via the existing strategy that @bkonyi has written about.  However, we will have even better support for it in the future.  No timeline is fixed, though.

**Comment by woozbooz** [comment]
Hello. At first sorry for my English.
I found this code in 

> **bkonyi**

 example (executing-dart-in-the-background-with-flutter-plugins-and-geofencing):
```
val channel = NotificationChannel(CHANNEL_ID, "Flutter Geofencing Plugin", NotificationManager.IMPORTANCE_LOW)
```
Did this (NotificationChannel) will work on Android 5/6/7?

**Comment by Hixie** [comment]
It would be helpful if we could split this bug into one per specific request rather than having a metabug. Metabugs tend to be overwhelming and not get traction as easily as specific concrete bugs.

**Comment by matthew-carroll** [comment]
I'll assign this to myself for the purpose of breaking it up and then un-assign for future work.

**Comment by matthew-carroll** [comment]
I've attempted to capture the specific action items from this ticket in a new meta bug that links to individual work items:

https://github.com/flutter/flutter/issues/32164

If there is a specific deliverable that is not listed in the new metabug, please add a comment and I'll update the bug description.

I'm going to close this ticket as it has been migrated elsewhere.

**Comment by yumnapratista24** [comment]
Kayak bias ngoding aja dah

**Comment by yumnapratista24** [comment]
> Kayak bias ngoding aja dah

wkwkwkwkwk

**Comment by Krunal79-flutter** [comment]
already solved at [here](https://icetutor.com/question/how-do-i-run-code-in-the-background-even-with-the-screen-off/)

**Comment by bemesquita** [comment]
@Krunal79-flutter no, its not solved. Customization like frequency only works for Android.

**Comment by z85510** [comment]
how can I read SMS when app isn't running and save it in database ?

**Comment by sandstrom99** [comment]
I'm currently working on a timer app. The state of the timer is currently stored and handled in dart/flutter. I'd like to show a notification natively to the user when the app is in the background and the timer is still running. This is currently easily done - but I'd also like to be able to pause/resume and stop the timer from the notification. Since I don't want to recreate my domain and data layer in Android and IOS, I'd like to reuse my already existing dart code. 

Im unsure whether this is currently possible? I hope this is the right place to voice my specific needs for background execution.
